swift/Arrow/Sources/Arrow/ArrowCExporter.swift

```diff
@@ -0,0 +1,134 @@
+        init(_ arrowType: ArrowType, name: String = "") throws {
+            self.arrowType = arrowType
+            self.arrowTypeName = try arrowType.cDataFormatId.cstring
+            self.name = name.cstring
+            super.init()
+        }
+    }
+
+    private class ExportArray: ExportData {
+        private let arrowData: ArrowData
+        private(set) var data = [UnsafeRawPointer?]()
+        private(set) var buffers: UnsafeMutablePointer<UnsafeRawPointer?>
+        init(_ arrowData: ArrowData) {
+            // keep a reference to the ArrowData
+            // obj so the memory doesn't get
+            // deallocated
+            self.arrowData = arrowData
+            for arrowBuffer in arrowData.buffers {
+                data.append(arrowBuffer.rawPointer)
+            }
+
+            self.buffers = UnsafeMutablePointer(mutating: data)
+            super.init()
+        }
+    }
+
+    private static var exportedData = [Int: ExportData]()
+    public init() {}
+
+    public func exportType(_ cSchema: inout ArrowC.ArrowSchema, arrowType: ArrowType, name: String = "") ->
+        Result<Bool, ArrowError> {
+        do {
+            let exportSchema = try ExportSchema(arrowType, name: name)
+            cSchema.format = exportSchema.arrowTypeName
+            cSchema.name = exportSchema.name
+            cSchema.private_data =
+                UnsafeMutableRawPointer(mutating: UnsafeRawPointer(bitPattern: exportSchema.id))
+            cSchema.release = {(data: UnsafeMutablePointer<ArrowC.ArrowSchema>?) in
+                let arraySchema = data!.pointee
+                let exportId = Int(bitPattern: arraySchema.private_data)
+                guard ArrowCExporter.exportedData[exportId] != nil else {
+                    fatalError("Export schema not found with id \(exportId)")
+                }
+
+                // the data associated with this exportSchema object
+                // which includes the C strings for the format and name
+                // be deallocated upon removal
+                ArrowCExporter.exportedData.removeValue(forKey: exportId)
+                ArrowC.ArrowSwiftClearReleaseSchema(data)
+            }
+        } catch {
+            return .failure(.unknownError("\(error)"))
+        }
+        return .success(true)
+    }
+
+    public func exportField(_ schema: inout ArrowC.ArrowSchema, field: ArrowField) ->
+        Result<Bool, ArrowError> {
+            return exportType(&schema, arrowType: field.type, name: field.name)
+    }
+
+    public func exportArray(_ cArray: inout ArrowC.ArrowArray, arrowData: ArrowData) {
+        let exportArray = ExportArray(arrowData)
+        cArray.buffers = exportArray.buffers
+        cArray.length = Int64(arrowData.length)
+        cArray.null_count = Int64(arrowData.nullCount)
+        cArray.n_buffers = Int64(arrowData.buffers.count)
+        cArray.n_children = 0
+        cArray.children = nil
+        cArray.dictionary = nil
```

Comment:
Hmm, is this for TODO later? Otherwise I think `n_children` should not be zero always? And also for `children` and `dictionary`.

Reply:
The Swift version of Arrow currently doesn't support having children (Struct type) or dictionaries.  I will add a comment mentioning this so it can be updated when those constructs are added.